procrastinate/aiopg_connector.py

```diff
@@ -66,45 +67,49 @@ def _wrap_json(self, arguments: Dict[str, Any]):
         }
 
     async def _get_connection(self) -> aiopg.Connection:
-        if self._connection and not self._connection.closed:
-            return self._connection
-
-        # tell aiopg not to register adapters for hstore & json by default, as
-        # those are registered at the module level and could overwrite previously
-        # defined adapters
-        kwargs = self._connection_parameters
-        kwargs.setdefault("enable_json", False)
-        kwargs.setdefault("enable_hstore", False)
-        kwargs.setdefault("enable_uuid", False)
-        conn = await aiopg.connect(**kwargs)
-        if self.json_loads:
-            psycopg2.extras.register_default_jsonb(conn.raw, loads=self.json_loads)
-        self._connection = conn
-        return conn
+        async with self.lock:
+            if self._connection and not self._connection.closed:
+                return self._connection
```

Comment:
This is vital that this part is inside the lock, but I wonder if we could add a fast path by copying this just before taking the lock.

The idea is that once the connection is created (and not closed), it's useless to wait on the lock. It's not even two possible to have two coroutine inside the `if` at the same time as there is no async op. So instead of waiting + taking the lock every time we get the connection, we could return immediately.

Reply:
:+1:

Reply:
Actually it does work. Here's what happens:

- coroutine 1 acquires the lock, and calls `aiopg.connect`
- coroutine 2 doesn't take the fast path because the connection is not yet open, so it blocks on the lock aquisition
- when coroutine 1 releases the lock, coroutine 2 acquires it and calls `aiopg.connect` **BUG**

Reply:
That's why I said "copying this" :) . We need the `if …: return …` both before & inside the lock.

Reply:
Oh ok, I had missed that part from your comment. Sorry. Sounds like an unnecessary optimization to me, but why not…

Reply:
I added the fast path.